File: lib/ruby/mango.rb
require "socket"
require "json"

module Mango
  module_function

  def socket_path
    sig = ENV["MANGO_INSTANCE_SIGNATURE"]
    return sig if sig

    raise "MANGO_INSTANCE_SIGNATURE is not set"
  end

  def call(cmd)
    sock = UNIXSocket.new(socket_path)
    sock.puts(cmd)
    line = sock.gets
    sock.close
    raise "connection closed without reply: #{cmd}" unless line

    JSON.parse(line)
  end

  def ensure_ok(reply)
    raise reply["error"] if reply["error"]

    reply
  end

  def retry(cmd, attempts = 50, delay = 0.1)
    attempts.times do
      begin
        return ensure_ok(call(cmd))
      rescue SystemCallError
        sleep delay
      end
    end
    raise "mango not reachable after #{attempts} attempts: #{cmd}"
  end

  def get(*spec)
    ensure_ok(call("get #{spec.join(" ")}"))
  end

  def set_option(key, value)
    ensure_ok(call("setoption #{key} #{value}"))
  end

  def dispatch(function, *args)
    ensure_ok(call("dispatch #{([function] + args).join(",")}"))
  end

  def unset_bind(mode, mods, keysym, family = "bind")
    ensure_ok(call("unset bind #{mode} #{mods} #{keysym} #{family}"))
  end

  def unset_rule(kind, spec)
    ensure_ok(call("unset #{kind} #{spec}"))
  end

  def watch(subject)
    sock = UNIXSocket.new(socket_path)
    sock.puts("watch #{subject}")
    while (line = sock.gets)
      next if line.strip.empty?

      event = JSON.parse(line)
      ensure_ok(event)
      yield event
    end
  ensure
    sock&.close
  end
end

File: lib/python/mango.py
import json
import os
import socket
import time


def socket_path():
    path = os.environ.get("MANGO_INSTANCE_SIGNATURE")
    if path:
        return path
    raise RuntimeError("MANGO_INSTANCE_SIGNATURE is not set")


def connect():
    sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
    sock.connect(socket_path())
    return sock


def call(cmd):
    sock = connect()
    try:
        sock.sendall((cmd + "\n").encode())
        line = sock.makefile("r").readline()
    finally:
        sock.close()
    if not line:
        raise RuntimeError(f"connection closed without reply: {cmd}")
    return json.loads(line)


def ensure(reply):
    if "error" in reply:
        raise RuntimeError(reply["error"])
    return reply


def retry(cmd, attempts=50, delay=0.1):
    for _ in range(attempts):
        try:
            return ensure(call(cmd))
        except OSError:
            time.sleep(delay)
    raise RuntimeError(f"mango not reachable after {attempts} attempts: {cmd}")


def get(*spec):
    return ensure(call("get " + " ".join(spec)))


def set_option(key, value):
    ensure(call(f"setoption {key} {value}"))


def dispatch(function, *args):
    ensure(call("dispatch " + ",".join((function,) + args)))


def unset_bind(mode, mods, keysym, family="bind"):
    ensure(call(f"unset bind {mode} {mods} {keysym} {family}"))


def unset_rule(kind, spec):
    ensure(call(f"unset {kind} {spec}"))


def watch(subject, on_event):
    sock = connect()
    try:
        sock.sendall(f"watch {subject}\n".encode())
        for line in sock.makefile("r"):
            if not line.strip():
                continue
            on_event(ensure(json.loads(line)))
    finally:
        sock.close()